src/test/java/io/datanerds/avropatch/PatchTest.java

```diff
@@ -4,36 +4,104 @@
-                .addEqualityGroup(new Patch(operations))
-                .addEqualityGroup(new Patch(ImmutableList.of(randomOperation()), headers))
-                .addEqualityGroup(new Patch(ImmutableList.of(randomOperation())))
+                .addEqualityGroup(new Patch<>(resource, operations, headers), new Patch<>(resource, operations, headers))
+                .addEqualityGroup(new Patch<>(resource, operations))
+                .addEqualityGroup(new Patch<>(resource, ImmutableList.of(randomOperation()),
+                        ImmutableMap.of("test 1", "bar", "test 2", uuid)))
+                .addEqualityGroup(new Patch<>(resource, ImmutableList.of(randomOperation())))
+                .addEqualityGroup(new Patch<>(resource, ImmutableList.of(randomOperation())))
                 .testEquals();
     }
 
+    @Test
+    public void getHeader() {
+        Patch<?> patch = new Patch<>(resource, operations, headers);
+        String stringHeader = patch.getHeader("test 1");
+        assertThat(stringHeader, is(equalTo("foo")));
+        UUID uuidHeader = patch.getHeader("test 2");
+        assertThat(uuidHeader, is(equalTo(uuid)));
+    }
+
+    @Test
+    public void isEmpty() {
+        assertTrue(new Patch<>(resource).isEmpty());
+        assertTrue(new Patch<>(resource, headers).isEmpty());
+        assertFalse(new Patch<>(resource, operations, headers).isEmpty());
+    }
+
+    @Test
+    public void size() {
+        assertThat(new Patch<>(resource).size(), is(equalTo(0)));
+        assertThat(new Patch<>(resource, operations, headers).size(), is(equalTo(operations.size())));
+    }
+
+    @Test
+    public void stream() {
+        assertThat(operations, is(equalTo(new Patch<>(resource, operations).stream().collect(Collectors.toList()))));
+    }
+
+    @Test
+    public void getter() {
+        Date timestamp = new Date();
+        Patch<UUID> patch = new Patch<>(resource, operations, headers, timestamp);
+        assertThat(patch.getResource(), is(equalTo(resource)));
+        assertThat(patch.getTimestamp(), is(equalTo(timestamp)));
+        for (int i = 0; i < operations.size(); i++) {
+            assertThat(patch.getOperation(i), is(equalTo(operations.get(i))));
+        }
+    }
+
+    @Test
+    public void getNotExistentHeader() {
+        String value = new Patch<>(resource, operations, headers).getHeader("test 7");
+        assertNull(value);
+    }
+
+    @Test
+    public void getHeaderWithIncompatibleType() {
+        Patch<UUID> patch = new Patch<>(resource, operations, headers);
+        assertTrue(patch.hasHeader("test 1"));
+
+        exception.expect(ClassCastException.class);
+        @SuppressWarnings("unused")
+        UUID value = patch.getHeader("test 1");
+    }
+
     @Test(expected = UnsupportedOperationException.class)
-    public void immutableOperations() {
-        new Patch(operations, headers).getOperations().add(randomOperation());
+    public void immutableHeadersRemove() {
```

Comment:
Lets be a bit more descriptive in the method names here. Something like : `ExceptionWhenRemovingImmutableHeaders`